src/main/utility/synchronization/thread_parking.rs

```diff
@@ -107,10 +107,19 @@ impl ThreadUnparker {
 
     /// Unpark the assigned thread.
     pub fn unpark(&self) {
-        // rust guarantees that everything that happens before the `unpark()` is visible on the
-        // other thread after returning from `park()`, so the `ready_flag` should always be visible
-        // as true after the other thread returns from `park()` (the store should not be reordered
-        // after the `unpark()`):
+        // TODO: Rust does not guarantee any synchronization between the thread that parks and the
+        // thread that unparks:
+        //
+        // https://doc.rust-lang.org/nightly/std/thread/fn.park.html#park-and-unpark
+        //
+        // > Notice that being unblocked does not imply any synchronization with someone that
+        // > unparked this thread, it could also be spurious. For example, it would be a valid, but
+        // > inefficient, implementation to make both park and unpark return immediately without doing
+        // > anything.
+        //
+        // There is no guarentee that the change to `ready_flag` will be visible after the parked
+        // thread resumes, which may lead to a deadlock. In practice, rust currently does use an
+        // atomic with release-acquire ordering, but this might be relaxed in the future.
         // https://github.com/rust-lang/rust/blob/21b246587c2687935bd6004ffa5dcc4f4dd6600d/library/std/src/sys_common/thread_parker/futex.rs#L21-L27
         self.ready_flag.store(true, Ordering::Release);
```

Comment:
Orthogonal to this PR, but I'm not sure I understand the `Release` orderings with no `Acquire`. Should `park` be using `Acquire` in the `success` ordering? Without that, I don't think the `ThreadParker` itself is providing synchronization?

I agree though that even so it wouldn't address the potential deadlock issue you're describing here.

Reply:
Yeah I think that was a mistake. Ideally the thread `park`/`unpark` would be providing synchronization (and technically does as an implementation detail), and then I think the `ready_flag` operations could just be `Relaxed`. But since the documentation states that `park`/`unpark` does not provide synchronization, the `compare_exchange` should use `Acquire` in the success path to get the synchronization (but still doesn't fix the potential deadlock).

I'll change the `compare_exchange` to use `Acquire`.